core/src/main/java/org/elasticsearch/discovery/zen/PublishClusterStateAction.java

```diff
@@ -85,21 +88,24 @@
     private final DiscoverySettings discoverySettings;
     private final ClusterName clusterName;
     private final PendingClusterStatesQueue pendingStatesQueue;
+    private final CustomPrototypeRegistry registry;
 
     public PublishClusterStateAction(
             Settings settings,
             TransportService transportService,
             Supplier<ClusterState> clusterStateSupplier,
             NewPendingClusterStateListener listener,
             DiscoverySettings discoverySettings,
-            ClusterName clusterName) {
+            ClusterName clusterName,
+            CustomPrototypeRegistry registry) {
```

Comment:
>  but do we need both the NamedWriteable in the stream

Yes, otherwise nodes that receive a new cluster state don't know how the read the custom cluster state parts. (in the synchronized block line 388 till 407)

> and the explicit registry when reading a diff? 

Ideally we wouldn't need to pass the registry to the diff, but with the way it reads the custom diff map, I didn't see another of doing that without breaking bwc. (see [here](https://github.com/elastic/elasticsearch/pull/21243/files#diff-d16ada5a8f54350c95215b709fdf2c88R760), [here](https://github.com/elastic/elasticsearch/pull/21243/files#diff-b103f001f952a0e6f2a4540d71593dd6R602) and [here](https://github.com/elastic/elasticsearch/pull/21243/files#diff-d6d141c41772a9088a29c2b838e5d8c4R642))